### ISSUE TITLE
Question: Advice for training/testing data selection

### ISSUE THREAD

**Issue by NowayIndustries** [inspirehep/magpie#86]
Hi all, I've been playing around with magpie for a week now. The results of my training sessions so far are not really what I was expecting (in regards to accuracy) and my first few attempts to fix this have failed so far. So I come to you for guidance, my apologies for the somewhat lengthy essay this issue has become.

**Some background information:** 
I have access to a large amount of news articles that have been labelled by human editors. These labels are/were not always applied very consistently from what I can see. Currently these editors are assisted by another automated system that attempts to predict labels for new texts (these can be overwritten by the editors), however this system is not very accurate. My current assignment is to implement magpie into this process which eventually should provide better suggestions for the editors.

The main hurdle in this (and I assume any usage of a deep learning tool) is training the model for our corpus. There are two main problems with this:
1. I am a novice in AI/ML/DL
2. The corpus is not in English

For the first hurdle there is no immediate fix, apart from me learning by trial and error & doing my own research before asking questions.

For the second hurdle I have created a modified version of magpie in which I have replaced the English stopwords with the Dutch ones. (I've also set a Dutch snowball stemmer as the default, but this does not appear to be used within the magpie model)

**The problem** so far with my trained models is that after the first epoch the values all appear to stabilize on low loss/high accuracy, but the prediction of labels (even on actual training data) is very low (<0.10 typically). There are occasional higher scores (I believe the highest I've seen is 0.5), but these seem to be the exception rather then the norm.

_For your viewing pleasure: a typical graph for my trained models_
![typical graph of trained model](http://i.imgur.com/6LZ8jGi.png)

**I believe the problem is due to** the way I select/'validate' the training/testing data I have loaded into my environment (before passing said data to magpie).
Currently the selection process is automatically split into an 80/20 division, the data feeding process attempts to grab testing data evenly spread from the entire dataset it has found in the database. (Some pre-selection takes place here; texts have to have labels, all texts have to be in Dutch and from a certain category.)
Validation is done by checking if a label occurs at least X times in the training/testing data, as well as some generic checks (no missing labels, or empty texts etc.)
The way this is currently done there is no assurance that the labels I am training for occur in the testing data as well, which I assume to be the main issue here. This is because testing and training data are validated separately with no cross referencing between the two.

**My proposed solution is** is to move the splitting of the data from the data-feeding process into the validation process, and unifying my currently split 'raw' training/testing data. Then when validating, I can split up my raw data into training and testing sets whilst enforcing:

- That a label occurs at least X times in the data.
- That a label has around Y%  of its total occurrences in the testing data set.

As far as I can tell, 80/20 is generally accepted as a good split ratio between training/testing data, although some situations may warrant a more equal split depending on the accuracy of and the amount of data available.

As for X, I currently have it said to the fairly arbitrary amount of 100. This reduces the raw amount of available labels from 3.755 to 288 when validated (occur more then a 100 times in the data) on a corpus containing 13.024 training and 3.259 testing texts. (The size of the corpus is from my test to see if the graph would change with a bigger test set, the validated set is 260 training and 59 testing files (when X is 100)).

**And now for the actual questions part of this question:**
1. Are there any other changes required for the use of magpie with another language? (Apart from designing a whole new neural network ;)
2. Am I even on the right track with regards to validation/selection of training data?
3. Is an 80/20 split indeed reasonable for creating a magpie corpus?
4. I think 100 is to high for X, I think something like 50 would be more reasonable; What kind of value should I be looking at (perhaps Z% of files should have the label)?

**Comment by NowayIndustries** [comment]
Again apologies for the length of the main post of this issue, I swear I didn't intend to write such a long post ;)

As we speak I am starting work on my proposed solution since I think this will be in any case superior to my current implementation. Should an even better alternative be proposed then I will of course be switching to that.

If it turns out that it is in fact very simple to adjust magpie to a different language I think it would be nice to document that somewhere, so you don't keep getting repeat questions. (I did a quick search on the issues here on the repo but I couldn't find anything that quite matched what I want; some advice on selecting/setting up a training corpus.)

**Comment by NowayIndustries** [comment]
A small update: My proposed solution has shown some promising results and I have thus started training a larger corpus (which will take aproximately 10 hours).

_The results of my smaller corpus actually show some variance_
![new smaller corpus results](http://i.imgur.com/hZ0oCgW.png)

**Comment by jstypka** [comment]
Hey, I'll try to give you a proper response when I get a hand on a proper computer. But in the meantime - could you provide some more information about your dataset? How many samples are there in the train /test set? How many labels are you trying to predict? How frequent are those labels (are there many moderately popular labels or a few massively popular labels and a long tail)? Does Magpie do a good job with predicting most popular labels but fail with less frequent ones?

**Comment by NowayIndustries** [comment]
My corpusses are based on news articles that were scraped by the organization I am trying to implement Magpie for, currently they are being grabbed from the europanu environment which only means that they are shown on a specific product the company makes. Also only Dutch texts are being selected.

The corpus is populated with raw data (txt and lab files) through my wrapper around Magpie (a simple rest API that can start the training as well). This data is placed into the raw-data folder. When a training session is started the API looks through the raw-data folder to find 'valid' data.

First I frequency count the different labels, afterwards a simple check is done to determine whether or not to train for a label: is the frequency over the minimum I have set?

Then for every label I want to train: I split the associated txt&lab files into training and test data, so it can be passed to magpie.

-----
Since my smaller corpus doesn't take nine hours to train I'll focus on that, the difference with the larger corpus being that it has less articles loaded into it, and the minimum frequency of labels to train for has been lowered to 20.

My smaller corpus (only articles from 1-1-2017 onwards) has 798 files of the raw data, after validation only 660 remain. These validated files are then split into 529 training and 127 testing files. These files result in the fitting of 133899 vectors (that is the validated files only). This corpus predicts for 29 labels, out of the 528 present (thus only 29 labels are assigned to more then 20 txt's).

These are the labels magpie is training for, and their ocurrence (the red line being the mininum frequency requirement): The most frequent label being 'europanu homepage' has 279 occurences, where the second most frequent 'beleid buitenland' has only 123 occurrences.
![smallcorpuslabels](http://i.imgur.com/j2h8ck2.png)
And here is magpie's output for the training of the small corpus:
[small-corpus-magpieoutput.txt](https://github.com/inspirehep/magpie/files/874942/small-corpus-magpieoutput.txt)
And the plot of those training values looks like this:
![smallcorpusplot](http://i.imgur.com/6Nho4Fe.png)
And a random txt&lab of files from the training set (.lab.txt because github doesnt like .lab):
[vkbbfdz6zizw.txt](https://github.com/inspirehep/magpie/files/874958/vkbbfdz6zizw.txt)
[vkbbfdz6zizw.lab.txt](https://github.com/inspirehep/magpie/files/874960/vkbbfdz6zizw.lab.txt)
The prediction for the linked txt (vkbbfdz6zizw.txt):
- "homepage Europa Nu": 0.38980045914649963
- "ANP Politiek 2" 0.2501259446144104
- "Beleid economie en geld" 0.2204069197177887
- "Nederland en Europa" 0.15696585178375244
-  "Beleid recht en veiligheid" 0.14330267906188965
- "Beleid buitenland" 0.1365947127342224
- "politieke monitor ANP" 0.10960877686738968
- "Europese Unie in cijfers" 0.08555332571268082
- (cut off for brevity)

The text is about the increased trust of consumers and producers in the eurozone's economy, so ignoring the first two labels (which are both products of the company iirc): 
the next label is 'policy economy and money' and 'Netherlands and Europe' after that. 

That seems pretty reasonable, only one other label assigned to the text is actually predicted for 'Europese unie in cijfers' which has gotten the fairly low rating of 0.08. I assume that is because it only occurs 26 times in the training data (of wich 6 or so should have been allocated as testing data aswell).

**Comment by jstypka** [comment]
Hey @NowayIndustries !

First, let me answer some of your questions:

1) Magpie should work out of the box for any language. I used it on both English and Polish and it worked fine. The stopwords and the stemmer as you noticed are not used in the model, they were used before I introduced `word2vec` into the model.
2 and 3) Yeah, it seems right. 80/20 is reasonable and you rightfully recognised that if a label lands in the test set, but not the training set that Magpie won't learn how to predict it. My suggestion would be to pick the X most popular labels and try to make the model work for them. Predicting exotic labels is very hard and Magpie sometime can learn a bias for the popular ones. It seems like you did that with the red line - 20 occurrences should be a decent threshold.

A thing that you might want to keep in mind - with Magpie, each time you're training two models actually. One is the `word2vec` model which trains language representation (you teach your computer speak Dutch ;-)) and then you train a neural net to perform classification.

If your task is not academic and you don't need to publish rigidly correct results, I would consider pretraining the `word2vec` model on a large corpus of Dutch language. The best case would be to use the whole corpus that you have, you can even enhance it with Dutch Wikipedia to create good word models. This training is fast and should take minutes.

For training the NN model, you definitely want to split the corpus though. The size that you mentioned 660 in total is quite small, I'm not sure how good results can be obtained with it but with the example you gave me, we can definitely assess that it learns something.

If this is only a subset - try to take a larger corpus and filter it so it contains only popular and not-exotic labels and train it on that. My very rough estimate would be a total corpus of size ~10k and 20-50 labels in it - this should produce decent results.

Let me know how it works!

**Comment by NowayIndustries** [comment]
Hi @jstypka 
Thanks for the quick response, my model for the large corpus is almost done training. This large corpus has 308 labels above the threshold of 20, with 12.352 training and 3.706 testing files.

If I recall correctly fitting the word2vec only took ~12 minutes or so (I did notice it's a single core process).

The reported values so far appear to be pretty stagnant, does this then imply that there should be little to no need to keep on training after the second epoch? I'll have to find a good way to actually use the callbacks to stop the training when stagnating (I don't suppose there's a simple function call for that?)
```
Fitted to 303437 vectors
Fitted to 329712 vectors
Fitted to 281848 vectors
Fitted to 300402 vectors
Fitted to 292352 vectors
Fitted to 289232 vectors
Fitted to 290572 vectors
Fitted to 321830 vectors
Fitted to 295073 vectors
Fitted to 297451 vectors
Fitted to 320514 vectors
Fitted to 344353 vectors
Fitted to 18483 vectors
Train on 12352 samples, validate on 3706 samples
Epoch 1/5
12352/12352 [==============================] - 2575s - loss: 0.0655 - acc: 0.9839 - val_loss: 0.0490 - val_acc: 0.9877
Epoch 2/5
12352/12352 [==============================] - 2419s - loss: 0.0454 - acc: 0.9887 - val_loss: 0.0445 - val_acc: 0.9884
Epoch 3/5
12352/12352 [==============================] - 2303s - loss: 0.0426 - acc: 0.9890 - val_loss: 0.0424 - val_acc: 0.9886
Epoch 4/5
12352/12352 [==============================] - 2361s - loss: 0.0411 - acc: 0.9891 - val_loss: 0.0411 - val_acc: 0.9885
Epoch 5/5
5120/12352 [===========>..................] - ETA: 1080s - loss: 0.0401 - acc: 0.9891
```
Once it's done I'll have to run some more predictions using test data as well as training data.

One more question: Is it safe to assume that the minimum occurrence count has a direct relationship to the certainty of prediction (so a higher minimum occurrence results in a higher prediction for a label)? 
I realize the answer is probably not as black and white as I would like it to be, but I will probably have to attempt to increase the scores that magpie is assigning to labels.

**Comment by jstypka** [comment]
The general rule of thumb is that the more frequent a label is, the more biased Magpie can be towards this label. If one category occurs in more than half of the texts, then I would expect Magpie to give it reasonably high scores to pretty much any input. There are probably some ways of tackling this (subsampling?), but I have never tried that.

There is an easy way to pass a callback to stop execution for an arbitrary condition. The train method has a parameter `callbacks` where you can pass a list of [keras callbacks](https://keras.io/callbacks/#earlystopping). The one that you're looking for is `EarlyStopping`, but consider also `ModelCheckpoint` if the training takes a long time.

Remember that you don't have to retrain `word2vec` model every time. You can train it once and pass it as a parameter later on, this'll save you a couple of minutes.

The general rule is that if validation loss (`val_loss`) is decreasing, it's still worthwhile to train longer. However if the training takes super long for you, then maybe the gain from additional epochs is not worth the wait. It's difficult to say, but I think you should try at least once to leave it running for long to see what is the minimal `val_loss` that it saturates on. Is it 0.04 or maybe 0.01?

Are you training the model on your laptop? Is it using only one core or all of them? If you have TensorFlow as the backend library, it should support multicore training by default.

**Comment by jstypka** [comment]
~300 labels for 15k files should be able to get a decent performance as far as my knowledge goes

**Comment by NowayIndustries** [comment]
Again thanks for all of your input, it would have taken me a lot longer to come to these solutions without it.

For my use-case I don't think the over representation of some labels in the training texts will be a problem, the extracted news articles and such are already put into a certain category/environment for the products the company makes. I will be able to setup different corpusses for these categories, and the over representation may even be an advantage. Magpie will mimic what our editors have done in the past and react accordingly. 

The earlystopping callback is exactly what I was looking for, I was already using the other callback for epoch ends to report the output of magpie through my API.

Not retraining (or even pre-training) a word2vec model is something to be considered, but part of the use-case (as it stands now) is to organically grow magpie's knowledge and have the ability to add new labels (which will be done by adding the required amount of data with said label of-course). Besides, 12 minutes on a ~40 minutes per epoch is not too bad. 

It will probably become a balance between run time of the training and the val_loss parameter, it will have to be included in the testing phase to see what sort of values make sense here.

I am training the model on a virtual server I SSH into in order to develop my API, the actual training of the network is multi-threaded and loads up all 12 cores I have available (For an average usage of 1100% CPU, which looks amazing in htop). The initialization of the `word2vec` model is only pinning one core on my machine, altough multi-processing it might be more hassle then it's worth.
I'm currently using Theano as keras back-end because I had some issues installing Tensorflow & magpie into a virtual environment when first setting up a small magpie test environment. I think that issue has since been resolved with the bumping of the version numbers in the requirements.txt. (I think specifically it was Keras that needed an update after Tensorflow went to its 1.0(?) release and swapped around a bunch of parameters)

I think it's safe to say my questions have been answered, I have marked this issue as closed. Thanks again for all the help and information!
